File: github-milestone/in.js
#!/usr/bin/env node

const fs = require('fs');
const path = require('path');
const fetch = require('node-fetch');

const outdir = process.argv[2];
const stdin = fs.readFileSync(0, 'utf-8');
const config = JSON.parse(stdin);

const username = config.source.username;
const password = config.source.password;
const repo = config.source.repository;

const endpoint = `http://api.github.com/repos/${repo}/milestones`;
const headers = {
  'Authorization': 'Basic ' + Buffer.from(username + ':' + password).toString('base64')
}

const writeFiles = dest => milestone => {
  const jsonFile = path.resolve(dest, 'milestone.json');
  const titleFile = path.resolve(dest, 'title.txt');
  const slugFile = path.resolve(dest, 'title-slug.txt');
  const numberFile = path.resolve(dest, 'number.txt');
  const idFile = path.resolve(dest, 'id.txt');

  fs.writeFileSync(jsonFile, JSON.stringify(milestone));
  fs.writeFileSync(titleFile, milestone.title);
  fs.writeFileSync(slugFile, milestone.title.replace(/(^[^a-z]{1})|([^a-z0-9 \-]+)|([^a-z0-9]+$)/g, '').replace(/[ \-]+/g, '-'));
  fs.writeFileSync(numberFile, milestone.number);
  fs.writeFileSync(idFile, milestone.id);

  return milestone;
};

const writeOut = result => {
  process.stdout.write(JSON.stringify(result));
};

if (config.source.mode === 'multiple') {
  const versions = config.version.ids.split(',');
  const request = fetch(endpoint, {headers})
    .then(response => response.json())
    .then(milestones => milestones.filter(milestone => versions.indexOf('' + milestone.number) > -1))
    .then(milestones => {
      milestones.forEach(milestone => {
        const dest = path.resolve(outdir, '' + milestone.number);
        fs.mkdirSync(dest);
        writeFiles(dest)(milestone);
      });
      return milestones;
    })
    .then(milestones => ({
      version: config.version,
      metadata: milestones.map(milestone => ({name: '' + milestone.number, value: milestone.title}))
    }))
    .then(writeOut)
  ;
} else {
  const version = config.version.id;

  fetch(`${endpoint}/${version}`, {headers})
    .then(response => response.json())
    .then(writeFiles(outdir))
    .then(milestone => ({
      version: {
        id: '' + milestone.number,
      },
      metadata: Object.entries(milestone)
        .filter(entry => ['string', 'number'].indexOf(typeof(entry[1])) > -1)
        .map(([name, value]) => ({name, value: '' + value}))
    }))
    .then(writeOut)
  ;
}

File: github-issue/in.py
#!/usr/bin/env python3

import requests
import json
import sys
import os
import datetime

def _in(instream):
    payload = json.load(instream)
    path = str(sys.argv[1])
    source = payload['source']
    token = source['token']
    repository = source['repository']
    version = [payload['version']]
    versionNumber = version[-1]['number']
    versionDate = version[-1]['modified']

    endpoint = "https://api.github.com/repos/" + repository + "/issues/" + versionNumber
    headers = {'Authorization': 'token ' + token}
    connection = requests.get(endpoint, headers=headers)
    now = datetime.datetime.now() - datetime.timedelta(seconds = 180)
    date = now.isoformat()

    if  versionDate >= date:
        with open(path + '/issue.json', 'w+') as issue:
            json.dump(connection.text, issue)
            issue.close()

        title = open(path + '/title.txt', 'w+')
        title.write(connection.json()['title'])
        title.read()
        title.close()

        numfile = open(path + '/number.txt', 'w+')
        numfile.write(str(connection.json()['number']))
        numfile.read()
        numfile.close()

        body = open(path + '/body.txt', 'w+')
        body.write(connection.json()['body'])
        body.read()
        body.close()

    else:
        print("Dates do not match", file=sys.stderr)
        sys.exit(1)

if __name__ == "__main__":
    print(json.dumps(_in(sys.stdin)))
